File: ios/joinery-kit/Sources/JoineryBillingKit/BillingStore.swift
import Foundation
import StoreKit
import JoineryKit

/// State for the billing screen: server catalog + summary, StoreKit products,
/// and the purchase/restore flows. Server-authoritative — every StoreKit
/// success is posted to `store/app_store_claim` before the UI reports it.
@MainActor
public final class BillingStore: ObservableObject {
    public enum Phase {
        case loading
        case failed(String)
        case loaded
    }

    @Published public private(set) var phase: Phase = .loading
    @Published public private(set) var catalog: BillingCatalog?
    @Published public private(set) var summary: BillingSummary?
    @Published public private(set) var storeProducts: [String: Product] = [:]
    @Published public private(set) var purchasing = false
    @Published public var actionError: String?
    @Published public var actionMessage: String?

    private let api: BillingAPI

    public init(api: BillingAPI) {
        self.api = api
    }

    public convenience init(client: APIClient) {
        self.init(api: BillingAPI(client: client))
    }

    public func initialLoad() async {
        phase = .loading
        await load()
    }

    public func load() async {
        do {
            async let catalogTask = api.catalog()
            async let summaryTask = api.summary()
            let (catalog, summary) = try await (catalogTask, summaryTask)
            self.catalog = catalog
            self.summary = summary
            if !catalog.plans.isEmpty {
                let ids = catalog.plans.map(\.storeProductID)
                let products = (try? await Product.products(for: ids)) ?? []
                storeProducts = Dictionary(uniqueKeysWithValues: products.map { ($0.id, $0) })
            } else {
                storeProducts = [:]
            }
            phase = .loaded
        } catch let error as JoineryAPIError {
            phase = .failed(error.displayMessage)
        } catch {
            phase = .failed("Could not load plans.")
        }
    }

    /// Run the StoreKit purchase sheet for a plan, then claim the signed
    /// transaction server-side. The tier is granted only when the claim
    /// succeeds.
    public func purchase(_ plan: BillingPlan) async {
        guard let catalog, !purchasing else { return }
        guard let product = storeProducts[plan.storeProductID] else {
            actionError = "This plan is not available right now."
            return
        }
        purchasing = true
        defer { purchasing = false }

        var options: Set<Product.PurchaseOption> = []
        if let token = UUID(uuidString: catalog.appAccountToken) {
            options.insert(.appAccountToken(token))
        }

        do {
            let result = try await product.purchase(options: options)
            switch result {
            case .success(let verification):
                guard case .verified(let transaction) = verification else {
                    actionError = "The purchase could not be verified."
                    return
                }
                let claim = try await api.claim(jws: verification.jwsRepresentation)
                await transaction.finish()
                actionMessage = claim.tier.map { "You're on \($0.name)." } ?? "Purchase complete."
                await load()
            case .userCancelled, .pending:
                break
            @unknown default:
                break
            }
        } catch let error as JoineryAPIError {
            actionError = error.displayMessage
        } catch {
            actionError = "The purchase could not be completed."
        }
    }

    /// Re-claim every current auto-renewable entitlement (restore purchases,
    /// reinstalls, new device). Claims are idempotent server-side.
    public func restorePurchases() async {
        guard !purchasing else { return }
        purchasing = true
        defer { purchasing = false }

        var claimed = 0
        var lastError: String?
        for await result in Transaction.currentEntitlements {
            guard case .verified(let transaction) = result,
                  transaction.productType == .autoRenewable else { continue }
            do {
                _ = try await api.claim(jws: result.jwsRepresentation)
                claimed += 1
            } catch let error as JoineryAPIError {
                lastError = error.displayMessage
            } catch {
                lastError = "A purchase could not be restored."
            }
        }

        if claimed > 0 {
            actionMessage = "Purchases restored."
            await load()
        } else if let lastError {
            actionError = lastError
        } else {
            actionError = "No purchases to restore."
        }
    }
}

File: android/joinery-android-billing/src/main/java/com/getjoinery/billing/BillingStore.kt
package com.getjoinery.billing

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.android.billingclient.api.ProductDetails
import com.getjoinery.android.JoineryApiError

sealed class BillingPhase {
    object Loading : BillingPhase()
    object Loaded : BillingPhase()
    data class Failed(val message: String) : BillingPhase()
}

internal fun displayMessage(e: Exception): String =
    (e as? JoineryApiError)?.displayMessage ?: (e.message ?: "Something went wrong.")

/**
 * Server-side state for the billing screen: catalog + summary loads and the
 * claim call. Play Billing state (ProductDetails, the purchase sheet) is fed
 * in by [PlayBillingConnector]; the tier is granted only when the server
 * accepts the claim — the screen reflects the server's view, not Play's.
 */
class BillingStore(val api: BillingApi) {
    var phase by mutableStateOf<BillingPhase>(BillingPhase.Loading)
        private set
    var catalog by mutableStateOf<BillingCatalog?>(null)
        private set
    var summary by mutableStateOf<BillingSummary?>(null)
        private set
    var productDetails by mutableStateOf<Map<String, ProductDetails>>(emptyMap())
        private set
    var purchasing by mutableStateOf(false)
    var actionError by mutableStateOf<String?>(null)
    var actionMessage by mutableStateOf<String?>(null)

    private var loadGeneration = 0

    suspend fun initialLoad() {
        phase = BillingPhase.Loading
        reload()
    }

    suspend fun reload() {
        loadGeneration += 1
        val generation = loadGeneration
        try {
            val newCatalog = api.catalog()
            val newSummary = api.summary()
            if (generation != loadGeneration) return
            catalog = newCatalog
            summary = newSummary
            phase = BillingPhase.Loaded
        } catch (e: Exception) {
            if (generation != loadGeneration) return
            phase = BillingPhase.Failed(displayMessage(e))
        }
    }

    fun onProductDetails(details: Map<String, ProductDetails>) {
        productDetails = details
    }

    /**
     * Claim purchase tokens server-side (post-purchase and restore). Returns
     * the number of successful claims; failures surface in [actionError].
     */
    suspend fun claimTokens(tokens: List<String>, packageName: String): Int {
        var claimed = 0
        var lastError: String? = null
        for (token in tokens) {
            try {
                val result = api.claim(token, packageName)
                claimed += 1
                actionMessage = result.tier?.let { "You're on ${it.name}." } ?: "Purchase complete."
            } catch (e: Exception) {
                lastError = displayMessage(e)
            }
        }
        if (claimed == 0 && lastError != null) actionError = lastError
        if (claimed > 0) reload()
        return claimed
    }

    fun clearActionError() { actionError = null }
    fun clearActionMessage() { actionMessage = null }
}
